### ISSUE TITLE
IE11 TypeError: Invalid character in header field name

### ISSUE THREAD

**Issue by cricketts** [okta/okta-auth-js#327]
Hello! We're running into a hard-to-troubleshoot issue. We're encounter the following error on Internet Explorer 11 consistently, but not in Chrome, Firefox, or Safari. For some reason it doesn't reproduce in IE11 Private browsing either.

The error that's getting thrown is the following:

```
../webpack:/root/okta/okta-auth-js/node_modules/cross-fetch/dist/browser-ponyfill.js:59:6t   n
```

```js
56      name = String(name);
57    }
58    if (/[^a-z0-9\-#$%&'*+.^_`\|~]/i.test(name)) {
59      throw new TypeError('Invalid character in header field name')
60    }
61    return name.toLowerCase()
62  }
```

<details>
<summary>Screenshot of entire stack trace</summary>

![image](https://user-images.githubusercontent.com/4120195/73900313-373de280-4844-11ea-9830-8ae8c9391ca1.png)

</details>

---

We started seeing this error from our users starting Feb 4th, 01:00 UTC at which point nothing of significant happened on our end (no deployments, changed to configuration, etc.).

The issue consistently reproduces against our prod application which uses a "production" developer Okta account, but it does not reproduce locally against our non-production developer Okta account.

Any advice would be appreciated.

**Comment by swiftone** [comment]
@cricketts - Thanks for the report.  A few questions for you to help us nail down what you might be experiencing.

- Are you using authJS directly, or also one of our front end SDKs (that use themselves use authjs)?
- Are you using `pkce: true`?
- What is the sequence of events that triggers the error, and when (e.g. does it happen before/after login, etc)

I'd also suggest reaching out to our support team at developers@okta.com and giving them the link to this issue - we'll coordinate efforts with them and you'll be able to share more details with them about your particular configuration.

**Comment by cricketts** [comment]
Thanks for the swift response! I'll reach out to the support team for further help troubleshooting.

> Are you using authJS directly, or also one of our front end SDKs (that use themselves use authjs)?

The only auth dependency we use on the client is `@okta/okta-auth-js@^2.12.0`. On the server we also use `@okta/jwt-verifier@^0.0.15` and `@okta/okta-sdk-nodejs@^2.0.1`.

> Are you using `pkce: true`?

If `pkce` is not enabled by default, then no we are not using it.

> What is the sequence of events that triggers the error, and when (e.g. does it happen before/after login, etc)

This error happens right after the user submits their credentials to login. I'll try to better pinpoint what library call triggers the error, but we call `oktaAuth.signIn()` and `oktaAuth.token.getWithoutPrompt()` during our sign in flow. Note our try/catches, so I think the error that's occurring is asynchronous:

```ts
const oktaAuth = new OktaAuth({...})

const authenticateWithSession = async (sessionToken?: string): Promise<[IToken, IAccessToken] | null> => {
  try {
    const options = {
      responseType: ['id_token', 'token'],
      scopes: ['openid', 'email', 'profile'],
      ...(sessionToken && {sessionToken}),
    }
    const token: [IToken, IAccessToken] = await oktaAuth.token
      .getWithoutPrompt(options)

    oktaAuth.tokenManager.add('appToken', token[0])
    oktaAuth.tokenManager.add('appAccessToken', token[1])
    return token
  } catch (err) {
    console.warn(err)
    return null
  }
}

export const authenticateWithCredentials = async (username, password) => {
  try {
    const tx = await oktaAuth.signIn({
      username,
      password,
    })
    return authenticateWithSession(tx.sessionToken)
  } catch (err) {
    bugsnagClient.notify(err, {severity: 'warning'})
    console.warn(err)
    return null
  }
}

**Comment by cricketts** [comment]
I can confirm that the error is thrown once the following function is called:

```ts
const tx = await oktaAuth.signIn({
  username,
  password,
})
```

**Comment by cricketts** [comment]
I was able to get a much simpler reproduction of the issue. I was able to confirm the request fails due to a slash in a request header attempting to be normalized. This same error is thrown when this code is run in the console:

```js
fetch('https://httpstat.us/200')
  .then(function(response) {
    return response.text()
  }).then(function(body) {
    console.log(body)
  })
```

This makes me think that this has to do with a `fetch` polyfill. We're continuing the investigation.

**Comment by cricketts** [comment]
[WalkMe](https://www.walkme.com/) seems to be the culprit here. We use this library to provide in-app support to our users. They load some code asynchronously, and our hypothesis is that they are polyfilling some code related to `fetch` which is causing this issue.

We disabled WalkMe in production and now users are able to login.

We're going to open a support issue with WalkMe because this seems likely unrelated to Okta at this point. If we find out any more information we'll post back here. Cheers!

**Comment by cricketts** [comment]
To close the loop here: we verified that WalkMe was introducing a polyfill for `fetch` that was incompatible with how Okta was authenticating. WalkMe fixed it by changing our settings to exclude a `menuVisualDesign` feature. We could validate that the feature was removed by running the following command in the console:

```js
_walkMe.getSiteConfig().Features.indexOf(‘menuVisualDesign’)
```

This fixed the issue and we're able to login successfully with IE11.
